File: output_20260318_091315/claude-opus/c/c/Struct/spouse.c
#include "stdlib.h"
#include "spouse.h"

struct person {
  char* name;
  struct person* spouse;
};

/*@ 
predicate person(struct person* p; pointer var) = 
  p != 0 &*& malloc_block_person(p) &*& p->name |-> _ &*& p->spouse |-> ?s &*& 
  (s == 0 || (s != 0 &*& s->spouse |-> p));
@*/

struct person *create_person()
  //@ requires true;
  //@ ensures person(result, _);
{
  struct person *p = malloc(sizeof(struct person));
  if(p == 0) abort();
  //@ p->name = 0;
  p->spouse = 0;
  
  return p;
}

void marry(struct person *this, struct person *other)
  //@ requires person(this, _) &*& person(other, _);
  //@ ensures person(this, _) &*& person(other, _) &*& this->spouse == other &*& other->spouse == this;
{
  this->spouse = other;
  other->spouse = this;
}

struct person* person_get_spouse(struct person* this)
  //@ requires person(this, _);
  //@ ensures person(this, _) &*& result == this->spouse;
{
  return this->spouse;
}

void divorce(struct person* this)
  //@ requires person(this, _);
  //@ ensures person(this, _) &*& this->spouse == 0;
{
  //@ struct person* s = this->spouse;
  this->spouse->spouse = 0;
  this->spouse = 0;
}

void die(struct person *this)
  //@ requires person(this, _);
  //@ ensures true;
{
  if(this->spouse != 0) {
    this->spouse->spouse = 0;
  }
  free(this); 
}

int main() 
  //@ requires true;
  //@ ensures true;
{
  struct person* alice = create_person();
  struct person* bob = create_person();
  struct person* eve = 0;
  marry(alice, bob);
  eve = create_person();
  divorce(bob);
  marry(bob, eve);
  die(alice);
  die(bob);
  die(eve);
  return 0;
}

File: output_at_5/claude-opus/sample_1/java/java/Linklist/Spouse.java
class Person {

    protected Person spouse;

    /*@ 
    predicate person(Object p) = p == this &*& true; 
    @*/

    //@ requires true;
    //@ ensures true;
    public void spouse_symm()
    {
    }

    //@ requires true;
    //@ ensures true;
    public Person()
    {
    }
    
    //@ requires true;
    //@ ensures true;
    public Person getSpouse()
    {
        return spouse;
    }
    
    //@ requires this != null &*& (other == null || other != null);
    //@ ensures true;
    protected void setSpouse(Person other)
    {
        spouse = other;
        other.spouse = this;
    }
    
    //@ requires this != null &*& spouse != null;
    //@ ensures true;
    protected void clearSpouse()
    {
        spouse.spouse = null;
        spouse = null;
    }
    
    //@ requires other != null;
    //@ ensures true;
    void marry(Person other)
    {
        other.setSpouse(this);
    }
    
    //@ requires spouse != null;
    //@ ensures true;
    void divorce()
    {
        spouse.clearSpouse();
    }

}

class Program {

    //@ requires true;
    //@ ensures true;
    public static void main(String[] args)
    {
        Person a = new Person();
        Person b = new Person();
        a.marry(b);
        b.divorce();
    }

}